### ISSUE TITLE
Mkdir errors parallel testing

### ISSUE THREAD

**Issue by dynasource** [yiisoft/yii2#9288]
In my parallel testing scenario (which could also potentially occur in production mode) 

```php
yii\base\Exception: Failed to create directory 'T:\wc6dev_test\assets\e87a0c9b\css': mkdir(): File exists
```
FileHelper::createDirectory can throw an error when it tries to create a folder which already exists.

https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseFileHelper.php#L465

What happens:
- it checks for a directory
- if not exist, it will try to create
- but when it is created in parallel already, it would throw an error.

The solution for this, is to re-check folder existance when an exception has occurred. If the folder is already created, throwing an error is not required anymore.

**Comment by dynasource** [comment]
I am still getting this error every day when I start my parallel testing enviroment

**Comment by dynasource** [comment]
```php
ArticleCest::update()
yii\base\Exception: Failed to create directory 'T:/lw5dev_test/runtime/cache': mkdir(): File exists
E:\xampp\htdocs\dmn\lw\5\dev\vendor\yiisoft\yii2\caching\FileCache.php:80
E:\xampp\htdocs\dmn\lw\5\dev\vendor\yiisoft\yii2\base\Object.php:107
E:\xampp\htdocs\dmn\lw\5\dev\vendor\yiisoft\yii2\di\Container.php:366
```

**Comment by dynasource** [comment]
@coredevelopers, can you please look at this:

In BaseFileHelper::createDirectory():
* an error is thrown for mkdir <--> when a folder already exists (when its just created in another request/thread)

Although this is logical from a sequential perspective in BaseFileHelper::createDirectory(), it isnt logical in itself and errors apply in practice. Usecases:
- testserver: parallel testing 
- production: many requests at the same time (not tested, but assumed)


```php
try {
            $result = mkdir($path, $mode);
            chmod($path, $mode);
        } catch (\Exception $e) {
            throw new \yii\base\Exception("Failed to create directory '$path': " . $e->getMessage(), $e->getCode(), $e);
        }
```

**Comment by samdark** [comment]
I think such behavior would be handy. I can't imagine a good use case when it won't fit. @yiisoft/core-developers thoughts?

**Comment by klimov-paul** [comment]
I am unsure this is correct way.
Error handling is created in the way in converts error into exceptions allow handling the problem.
The main concern is directory craetion is not the only thing, which may break during parallel execution.
For example `FileHelper::copyDirectory()` or `FileHelper::removeDirectory()` may also cause similar error in parallel execution.

I assume this issue is about asset pubishing. We already have similar request at #3214, which is about mutex usage.

Still, the fix can be made, but it should be considered carefully.

**Comment by dynasource** [comment]
Four solutions to choose from:
# fix it in createDirectory with a is_dir
# fix it in createDirectory with a @mkdir
# fix it with a central mkdir function
# fix it with mutex locks (if possible)

In Filehelper, I see this mkdir occurring only once. In a lot of Yii Framework files, I see the usage of @mkdir.

I would choose from solution 1:
* It isnt that important to give a own function like in 3
* Its important enough to not ignore other mkdir errors (like permission errors)
* It wouldnt have to be that complex to go for locking. 

The way I see it, is that the mkdir fuction of PHP is incomplete as it doesnt have an option to ignore the 'dir exists' option.

**Comment by samdark** [comment]
Yes. It's the best solution from what you've proposed. But I still wonder if there are cases when you need that error or want custom handling for it.

**Comment by klimov-paul** [comment]
>  But I still wonder if there are cases when you need that error or want custom handling for it.

Unlikely. The result method should look like following:

```php
public static function createDirectory($path, $mode = 0775, $recursive = true)
    {
        if (is_dir($path)) {
            return true;
        }
        $parentDir = dirname($path);
        if ($recursive && !is_dir($parentDir)) {
            static::createDirectory($parentDir, $mode, true);
        }
        try {
            $result = mkdir($path, $mode);
        } catch (\Exception $e) {
            if (!is_dir($path)) {
                throw new \yii\base\Exception("Failed to create directory '$path': " . $e->getMessage(), $e->getCode(), $e);
            }
        }
        try {
            chmod($path, $mode);
        } catch (\Exception $e) {
            throw new \yii\base\Exception("Failed to create directory '$path': " . $e->getMessage(), $e->getCode(), $e);
        }

        return $result;
    }
```

The idea is: if there is a error while creating a directory and directory exists in result - it is not error at all.

**Comment by dynasource** [comment]
@klimov-paul: from a logical point of view I would separate these two. Two remarks:
* you are returning a false $result in your example when the dirs exists
* the chmod exceptions deserves its own message  'Failed to chmod'

**Comment by samdark** [comment]
What if it exists but not chmodded correctly?

**Comment by klimov-paul** [comment]
> you are returning a false $result in your example when the dirs exists
> chmod exceptions deserves its own message 'Failed to chmod'

Sure: I have edited code above.

**Comment by dynasource** [comment]
The return could be simplyfied, as the $result of the mkdir function is no rocket science, its just a boolean http://php.net/manual/en/function.mkdir.php

I would choose for:
```php 
public static function createDirectory($path, $mode = 0775, $recursive = true)
    {
        if (is_dir($path)) {
            return true;
        }
        $parentDir = dirname($path);
        if ($recursive && !is_dir($parentDir)) {
            static::createDirectory($parentDir, $mode, true);
        }
        try {
            mkdir($path, $mode);
        } catch (\Exception $e) {
            if (!is_dir($path)) {
                throw new \yii\base\Exception("Failed to create directory '$path': " . $e->getMessage(), $e->getCode(), $e);
            } 
        }
        try {
            chmod($path, $mode);
        } catch (\Exception $e) {
            throw new \yii\base\Exception("Failed to change permissions for directory '$path': " . $e->getMessage(), $e->getCode(), $e);
        }

        return true;
    }

```

**Comment by dynasource** [comment]
still this function confuses me, ie:
* why Chmod a function when $result is false.

**Comment by dynasource** [comment]
Taken into account of that:
* its no use to chmod in case of a returning false of mkdir
* exceptions can occur for chmod due to parallel activities

wouldnt this version suffice:

```php
    public static function createDirectory($path, $mode = 0775, $recursive = true)
    {
        if (is_dir($path)) {
            return true;
        }
        $parentDir = dirname($path);
        if ($recursive && !is_dir($parentDir)) {
            static::createDirectory($parentDir, $mode, true);
        }
        try {
            if(!mkdir($path, $mode)){
                return false;
            }
        } catch (\Exception $e) {
            if (!is_dir($path)) {
                throw new \yii\base\Exception("Failed to create directory '$path': " . $e->getMessage(), $e->getCode(), $e);
            } 
        }
        try {
            return chmod($path, $mode);
        } catch (\Exception $e) {
            throw new \yii\base\Exception("Failed to change permissions for directory '$path': " . $e->getMessage(), $e->getCode(), $e);
        }

    }  
```

**Comment by dynasource** [comment]
this last version looks ok to me. Should I create a PR? (I would love to get rid of these mkdir errors)

**Comment by samdark** [comment]
Yes.

**Comment by samdark** [comment]
Merged: https://github.com/yiisoft/yii2/commit/4d21c48bc42a3ada4ec40e645de9f58a42f598f7

Thanks!

**Comment by doublehops** [comment]
I am also experiencing this problem in my development system. I am running Yii 2.0.6. Has this been resolved? It's hard to tell from reading these comments. Alternatively, is there a workaround for now?

**Comment by samdark** [comment]
Yes, it was resolved. If you have an issue, report separately with details.

**Comment by cebe** [comment]
@doublehops it will be resolved in 2.0.7 which is not yet released.

**Comment by mashhur** [comment]
Hi guys!
I am having a same problem on my mac (yosemite):
It throws an exception of  "Failed to create directory "/xxxx": mkdir(): Permission denied"
public static function createDirectory($path, $mode = 0775, $recursive = true)
    {
		echo("<br>");
		echo("Path: " . $path);
		echo("<br>");
        if (is_dir($path)) {
            return true;
        }
        $parentDir = dirname($path);
		echo("Parent Dir: " . $parentDir);
        // recurse if parent dir does not exist and we are not at the root of the file system.
        if ($recursive && !is_dir($parentDir) && $parentDir !== $path) {
            static::createDirectory($parentDir, $mode, true);
        }
        try {
            if (!mkdir($path, $mode)) {
                return false;
            }
        } catch (\Exception $e) {
            if (!is_dir($path)) {// https://github.com/yiisoft/yii2/issues/9288
                throw new \yii\base\Exception("Failed to create directory \"$path\": " . $e->getMessage(), $e->getCode(), $e);
            }
        }
        try {
            return chmod($path, $mode);
        } catch (\Exception $e) {
            throw new \yii\base\Exception("Failed to change permissions for directory \"$path\": " . $e->getMessage(), $e->getCode(), $e);
        }
    }

**Comment by SilverFire** [comment]
It is related to permissions in your FS. RTFM, please: http://linuxcommand.org/lts0070.php

**Comment by mashhur** [comment]
Yeap but what im not figuring out is:
My web app is in webroot/A/B but this .php script creating a folder in system root folder ( / ).

**Comment by SilverFire** [comment]
Please, create a separate issue.
